cpp/src/spatial/linestring_distance.cu

```diff
@@ -0,0 +1,356 @@
+ * @p num_points is returned.
+ */
+template <typename OffsetIterator>
+inline cudf::size_type __device__
+endpoint_index_of_linestring(cudf::size_type const& linestring_idx,
+                             OffsetIterator const& linestring_offsets_begin,
+                             cudf::size_type const& num_linestrings,
+                             cudf::size_type const& num_points)
+{
+  return (linestring_idx == (num_linestrings - 1)
+            ? (num_points)
+            : *(linestring_offsets_begin + linestring_idx + 1)) -
+         1;
+}
+
+/**
+ * @brief Computes shortest distance between @p C and segment @p A @p B
+ */
+template <typename T>
+double __device__ point_to_segment_distance(coord_2d<T> const& C,
+                                            coord_2d<T> const& A,
+                                            coord_2d<T> const& B)
+{
+  double L_squared = (A.x - B.x) * (A.x - B.x) + (A.y - B.y) * (A.y - B.y);
+  if (L_squared == 0) { return hypot(C.x - A.x, C.y - A.y); }
+  double r = ((C.x - A.x) * (B.x - A.x) + (C.y - A.y) * (B.y - A.y)) / L_squared;
+  if (r <= 0 or r >= 1) {
+    return std::min(hypot(C.x - A.x, C.y - A.y), hypot(C.x - B.x, C.y - B.y));
+  }
+  double Px = A.x + r * (B.x - A.x);
+  double Py = A.y + r * (B.y - A.y);
+  return hypot(C.x - Px, C.y - Py);
+}
+
+/**
+ * @brief Computes shortest distance between two segments that doesn't intersect.
+ */
+template <typename T>
+double __device__ segment_distance_no_intersect_or_collinear(coord_2d<T> const& A,
+                                                             coord_2d<T> const& B,
+                                                             coord_2d<T> const& C,
+                                                             coord_2d<T> const& D)
+{
+  return std::min(std::min(point_to_segment_distance(A, C, D), point_to_segment_distance(B, C, D)),
+                  std::min(point_to_segment_distance(C, A, B), point_to_segment_distance(D, A, B)));
+}
+
+/**
+ * @brief Computes shortest distance between two segments.
+ *
+ * If two segment intersects, distance is 0. Otherwise compute the shortest point
+ * to segment distance.
+ */
+template <typename T>
+double __device__ segment_distance(coord_2d<T> const& A,
+                                   coord_2d<T> const& B,
+                                   coord_2d<T> const& C,
+                                   coord_2d<T> const& D)
+{
+  double r_denom = (B.x - A.x) * (D.y - C.y) - (B.y - A.y) * (D.x - C.x);
+  double r_numer = (A.y - C.y) * (D.x - C.x) - (A.x - C.x) * (D.y - C.y);
+  if (r_denom == 0) {
+    // Segments parallel or collinear
+    return segment_distance_no_intersect_or_collinear(A, B, C, D);
+  }
+  double r = r_numer / r_denom;
+  double s = ((A.y - C.y) * (B.x - A.x) - (A.x - C.x) * (B.y - A.y)) /
+             ((B.x - A.x) * (D.y - C.y) - (B.y - A.y) * (D.x - C.x));
+  if (r >= 0 and r <= 1 and s >= 0 and s <= 1) { return 0.0; }
```

Comment:
> if (r_denom == 0) { // should we be using an approximate equals or < epsilon to avoid division by a very small number?

How do you usually pick epsilon when the type is templated?

Reply:
> auto s = cross(ab, ac) / cross(ab, cd);

I noticed in the other branch that `cross(ac, cd)` is recomputed here.